tests/rptest/tests/workload_upgrade_runner_test.py

```diff
@@ -0,0 +1,290 @@
+            status_progress = {
+                w: progress_lambda(w, version_param)
+                for w in to_check_list
+            }
+            for w, state in status_progress.items():
+                if state == PWorkload.DONE:
+                    self.logger.info(
+                        f"{w.get_workload_name()} {str_update_kind}")
+                    to_check_list.remove(w)
+
+            if delay := 1 - (time.time() - start_time) > 0:
+                # ensure that checks are not performed too fast, by requesting a delay of 1 second
+                time.sleep(delay)
+
+    def cluster_version(self) -> int:
+        return Admin(self.redpanda).get_features()['cluster_version']
+
+    @cluster(num_nodes=4)
+    def test_workloads_through_releases(self):
+        # this callback will be called between each upgrade, in a mixed version state
+        def mid_upgrade_check(raw_versions: dict[Any, RedpandaVersion]):
+            rp_versions = {
+                k: expand_version(self.installer, v)
+                for k, v in raw_versions.items()
+            }
+            next_version = max(rp_versions.values())
+            # check only workload that are active and that can operate with next_version
+            to_check_workloads = [
+                w for w in self.adapted_workloads
+                if w.state == WorkloadAdapter.STARTED
+                and next_version <= w.get_latest_applicable_release()
+            ]
+            self._check_workload_list(to_check_list=to_check_workloads,
+                                      version_param=rp_versions,
+                                      partial_update=True)
+
+        # upgrade loop: for each version
+        for current_version in self.upgrade_through_versions(
+                self.upgrade_steps,
+                already_running=False,
+                mid_upgrade_check=mid_upgrade_check):
+            current_version = expand_version(self.installer, current_version)
+            # setup workload that could start at current_version
+            for w in self.adapted_workloads:
+                if w.state == WorkloadAdapter.NOT_STARTED and current_version >= w.get_earliest_applicable_release(
+                ):
+                    self.logger.info(f"setup {w.get_workload_name()}")
+                    w.begin()  # this will set in a STARTED state
+
+            # run checks on all the started workload.
+            # each check could take multiple runs, so loop on a list of it until exhaustion
+            self._check_workload_list(to_check_list= \
+                                      [w for w in self.adapted_workloads if w.state == WorkloadAdapter.STARTED],
+                                      version_param=current_version)
+
+            # stop workload that can't operate with next_version
+            for w in self.adapted_workloads:
+                if w.state == WorkloadAdapter.STARTED and current_version == w.get_latest_applicable_release(
```

Comment:
The condition uses `current_version`, while we've not yet progressed to the next version. I'm guessing we should use the first version greater than it from `self.upgrade_steps`.